gtags/gtags.1
.\"
.\" Copyright (c) 1996, 1997, 1998, 1999
.\"             Shigio Yamaguchi. All rights reserved.
.\" Copyright (c) 1999, 2000, 2001
.\"             Tama Communications Corporation. All rights reserved.
.\"
.\" This file is part of GNU GLOBAL.
.\"
.\" GNU GLOBAL is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2, or (at your option)
.\" any later version.
.\"
.\" GNU GLOBAL is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.Dd July 20, 2001
.Dt GTAGS 1
.Os FreeBSD
.Sh NAME
.Nm gtags
.Nd create tag files for global.
.Sh SYNOPSIS
.Nm gtags
.Op Fl c
.Op Fl i
.Op Fl I
.Op Fl o
.Op Fl w
.Op Fl v
.Op Ar dbpath
.Sh DESCRIPTION
.Nm Gtags
makes GTAGS, GRTAGS, GSYMS and GPATH file for
.Xr global 1 .
.Nm Gtags
trace subdirectories, read source files,
locate symbols and save the information into tag files in current directory.
You should execute this command at the root of the source tree.
If your source directory is on a read only device like CDROM, specify
.Ar dbpath
of the directory on which make tag files.
.Pp
C, C++, yacc, java and assembler source files are supported.
.Pp
.Bl -tag -width Ds -compact
.It Fl c
.It Fl -compact
Make tag files with compact format.
.It Fl -config Ar name
Show the value of config variable
.Ar name .
If
.Ar name
is not specified then show whole of config entry.
.It Fl -gtagsconf Ar file
Load user's configuration from
.Ar file .
.It Fl i
.It Fl -incremental
Update tag files incrementally. You had better use
.Xr global 1
with the -u option.
.It Fl I
.It Fl -idutils
Make index files for id-utils(1).
.It Fl o
.It Fl -omit-gsyms
Suppress making GSYMS file.
Use this option if you don't use -s option of
.Xr global 1 .
.It Fl v
.It Fl -verbose
Verbose mode.
.It Fl w
.It Fl -warning
Print warning messages.
.El
.Pp
Files whose names end in
.Nm \&.c
or
.Nm \&.h
are assumed to be C
source files and are searched for C style routine and macro definitions.
Files whose names end in
.Nm \&.c++
.Nm \&.cc
.Nm \&.cpp
.Nm \&.cxx
.Nm \&.hxx
.Nm \&.C
.Nm \&.H
are assumed to be C++
source files.
Files whose names end in
.Nm \&.y
are assumed to be
.Tn YACC
source files.
Files whose names end in
.Nm \&.java
are assumed to be Java source files.
Files whose names end in
.Nm \&.s
or
.Nm \&.S
are assumed to be Assembler
source files.
Other files are searched for C style definitions.
.Sh FILES
.Bl -tag -width tags -compact
.It Pa GTAGS
Tag file for function definitions.
.It Pa GRTAGS
Tag file for function references.
.It Pa GSYMS
Tag file for other symbols.
.It Pa GPATH
Tag file for path of source files.
.It Pa /etc/gtags.conf $HOME/.globalrc
Configuration file.
.El
.Sh ENVIRONMENT
The following environment variables affect the execution of htags.
.Pp
.Bl -tag -width indent
.It Ev GTAGSCONF
If this variable is set, its value is used as the configuration file.
The default is $HOME/.globalrc.
.It Ev GTAGSLABEL
If this variable is set, its value is used as the label of configuration file.
The default is 'default'.
.El
.Sh EXAMPLES
  % ls -F
  Makefile      src/    lib/
  % gtags -v
  % global -x main
  main              10 src/main.c  main (argc, argv) {
.Sh DIAGNOSTICS
.Nm Gtags
exits with a non 0 value if an error occurred, 0 otherwise.
.Sh SEE ALSO
.Xr btreeop 1 ,
.Xr gctags 1 ,
.Xr global 1 ,
.Xr gozilla 1 ,
.Xr htags 1 ,
.Xr vi 1 .
.Sh BUG
GTAGS, GRTAGS and GSYMS are very large.
In advance of using this command, check the space of your disk.
.Pp
Assembler support is far from complete.  It extracts only ENTRY()
and ALTENTRY() from source file. Probably valid only for FreeBSD and Linux
kernel source.
.Pp
There is no concurrency control about tag files.
.Pp
Symbols in assembler source files are not extracted for GSYMS.
.Sh AUTHORS
Shigio Yamaguchi
.Sh HISTORY
The
.Nm
command appeared in FreeBSD 2.2.2.


NEWS
___________________________________
|      |  |  |     |  _  |     |  |
|  |___|  |  |  |  |    _|  |  |  |    GNU GLOBAL source code tagging system
|  |   |  |  |  |  |     |     |  |
|  ~~  |   ~~|     |  ~  |  |  |   ~~|          for all hackers.
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 Copyright (c) 2000-2016 Tama Communications Corporation

 This file is free software; as a special exception the author gives
 unlimited permission to copy and/or distribute it, with or without
 modifications, as long as this notice is preserved.

 This program is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY, to the extent permitted by law; without even the
 implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

		----------------------------------

GNU GLOBAL NEWS - User visible changes.
At least, you should search '[INCOMPATIBLE CHANGES]' for each version.

Version 6.5.7 - 15 May 2017

[CHANGES]
o global: Now global(1) accept a file as --nearness option.
  If it is a file, tags in the file are shown first.
o global: Added aliases for MAKEOBJDIR and MAKEOBJDIRPREFIX.
  GTAGSOBJDIR: alias for MAKEOBJDIR
  GTAGSOBJDIRPREFIX: alias for MAKEOBJDIRPREFIX
  These aliases are given higher priority than the originals.
o global: New command --print=<name>.
  <name> may be one of: root, dbpath or conf.
  This is a generic version of the -p (--print-dbpath) command.
	'global -p' is equal to 'global --print=dbpath'
	'global -pr' is equal to 'global --print=root'
o gtags: Updated the built-in PHP parser to support namespaces and traits,
  which were introduced in PHP5. The following keywords have been removed:
	- failure: not a keyword
	- header: normal built-in function
	- is_array: normal built-in function
	- is_set: typo -> isset
	- printf: normal built-in function
	- setcookie: normal built-in function
	- success: not a keyword

[FIXED BUGS]
o globash: Echo command with -n option didn't work on some OS. Now it works.
o global: The --nearness option without -x option brought segmentation fault.
	$ global --nearness=. main
	Segmentation fault: 11
  Now it works.
o global: --from-here with -i option did not ignore case of function name
  in the context. Now it works.
	Input:
	[main.c]
	1: main() {
	2:         func();
	3:         Func();
	4: }
	5: func() {}
	6: Func() {}

	$ global -xi --from-here=6:main.c func
	(global-6.5.6)
	Func                6 main.c           Func() {}
	func                5 main.c           func() {}
	(global-6.5.7)
	Func                3 main.c            Func();
	func                2 main.c            func();

[INCOMPATIBLE CHANGES]
o gtags.conf: Added new candidate path.
        (1) --gtagsconf file
        (2) $GTAGSCONF
        (3) [project root]/gtags.conf
        (4) [project root]/[objdir]/gtags.conf       <- [NEW]
        (5) $HOME/.globalrc
        (6) /etc/gtags.conf
        (7) [sysconfdir]/gtags.conf

        The default of [objdir] is 'obj'. You can change it by
        the environment variables (GTAGSOBJDIR or MAKEOBJDIR).

Version 6.5.6 - December 19 2016

[CHANGES]
New facilities:
o htags-server: New --retry[=n] option.
  If the port is already in use, retry n times with incrementing the port number.
  The default of n is 20.
o htags: Changed the format of function header (--show-position) to make copying
  text easier.
o geco.rc: Added 'fzf' as a selector candidate.
o gtags: New configuration variable 'gtags_hook'.
  gtags_hook=<command line>
	Specify a command line which should be executed at the beginning of gtags(1).
	Leading "./" in any path is always means the project root directory, since
	gtags(1) is always invoked there.
  GTAGS_COMMANDLINE environment variable (read only)
	You can get the effective command line of gtags(1) from the hook. It includes
	both $GTAGS_OPTIONS and real arguments.

   [Usage]
   You can update 'gtags.files' before gtags(1) read it.

	[gtags.conf]
	+----------------------------------------
	|...
	|:gtags_hook=find src lib -print >gtags.files:

   You can refer the effective arguments of gtags(1) from the hook using
   environment variable GTAGS_COMMANDLINE.

	[gtags.conf]
	+----------------------------------------
	|...
	|:gtags_hook=./gen.sh:\
	|:GTAGS_OPTIONS=-c:\

	[gen.sh]
	+----------------------------------------
	|#!/bin/sh
	|echo ">>> $GTAGS_COMMANDLINE"		# show effective command line

	$ gtags -O
	>>> gtags -c -O
	$ _

[INCOMPATIBLE CHANGES]
o htags: Now --cvsweb option always insert 'view=log' to the generated URLs.
  Because it seems to be almost always a desirable specification.

[FIXED BUGS]
o gtags: Gtags often aborts with a message "buffer overflow. strlimcpy" when
  it encounters a long token (> 152 bytes). Now gtags always ignores it with
  a message "symbol name is too long. (Ignored)".
o htags: Old packages included two CGI scripts (completion.cgi, global.cgi)
  generated in the release manager's machine by mistake. They have some literal
  path like '/opt/local/bin/perl' which works only with MacPorts.
  Now, they are generated in the target (your) machine.

Version 6.5.5 - Sep 21 2016

This is a bug fix release.

[FIXED BUGS]
o exuberant-ctags.la: The path name of ctags command couldn't changed.
  Now, it can be changed by configuration variable 'ctagscom'.
o Tag files got corrupted when the built-in C++ parser gave up parsing and aborted.
  Now, it is avoided.

o Some inappropriate error messages were changed.

  Error1: NULL file (file size == 0)

	$ cp /dev/null GTAGS
	$ global -x main

	(GLOBAL-6.5.4)	global: /tmp/test/GTAGS seems older format. Please remake tag files.
	(GLOBAL-6.5.5)	global: /tmp/test/GTAGS seems to be corrupted.

  Error2: Corrupted file

	$ echo aaa >GTAGS
	$ global -x main

	(GLOBAL-6.5.4)	global: GTAGS not found.
	(GLOBAL-6.5.5)	global: /tmp/test/GTAGS seems to be corrupted.

Version 6.5.4 - Mar 27 2016

This is a bug fix release.

[CHANGES]
Changed:
o gtags.conf: The entries which include '.xml' were commented out,

gtags.pl
#
# Gtags.pl --- Global facility for Nvi-1.81
#
# Copyright (c) 2001, 2002
#             Tama Communications Corporation. All rights reserved.
#
# This file is part of GNU GLOBAL.
#
# GNU GLOBAL is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# GNU GLOBAL is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
# usage:
#
#	nvi editor			command line
#	-----------------------------------------------------------
#	:perl tag qw(main)		global -x main
#	:perl tag qw(-T main)		global -Tx main
#	:perl tag qw(-r main)		global -rx main
#	:perl tag qw(-sl main)		global -slx main
#	:perl tag qw(-gi main)		global -gix main
#	:perl tag qw(-I main)		global -Ix main
#	:perl tag qw(-P file)		global -Px file
#	:perl tag qw(-f %)		global -fx <current file>
#	:perl gozilla			gozilla +<current line> <current file>
#
# Please type ':perl help[ENTER]' for help.
#
# Suggested .nexrc: (If you have gtags.pl in ${HOME}/perl.)
#
# perl use lib "$ENV{HOME}/perl"
# perl require 'gtags.pl'
# map ^P :tagprev^M
# map ^N :tagnext^M
# map ^] :perl tag^M
# map ^G :perl gozilla^M
# ab gtag perl tag qw(
# ab gta perl tag qw(
# ab gt perl tag qw(
# ab gozill perl gozilla
# ab gozil perl gozilla
# ab gozi perl gozilla
# ab goz perl gozilla

#
# You need this version of GLOBAL.
#
$support_version = 4.3;
#
# command existent check.
#
$w32 = ($^O =~ /^(ms)?(dos|win(32|nt))/i) ? 1 : 0;
$pathsep = ($w32) ? ';' : ':';
$command = '';
foreach (split(/$pathsep/, $ENV{'PATH'})) {
    if (-x "$_/global") {
        $command = "$_/global";
        last;
    }
}
#
# version check of global(1).
#
$notfound = '';
if ($command) {
    open(TAGS, "$command --version |");
    $_ = <TAGS>;
    chop($_);
    $version = $_;
    close(TAGS);
    if ($? != 0 || $version < $support_version) {
        $notfound = "Your global(1) seems to be older version.";
    }
} else {
    $notfound = "Global(1) not found.";
}
if ($notfound) {
    $notfound .= " You need GLOBAL-${support_version} or the later.";
}

sub main::tag {
    my $tagq, $tag, $flag;

    if ($notfound) {
        $curscr->Msg($notfound);
	return;
    }
    $flag = '';
    while ($_[0] =~ /^-\w+$/) {
        if ($_[0] =~ /^-([fgIPilrsTx]+)$/) {
            $flag .= $1;
            shift;
        } else {
            $curscr->Msg("'$_[0]' not acceptable. Please type ':perl help[ENTER]'.");
            return;
        }
    }
    if ($flag =~ /r/ && $flag =~ /s/) {
        $curscr->Msg("both of -s and -r are not allowed.");
        return;
    }
    if ($_[0]) {
        $tag = $_[0];
	#
	# replace '%' with current file name.
	#
	if ($tag =~ /%/) {
            $path = $curscr->GetFileName();
	    $tag =~ s/%/$path/;
	} else {
	}
    } else {
        #
        # get current position.
        #
        my($lineno, $column) = $curscr->GetCursor();
        my($line) = $curscr->GetLine($lineno);
	my($length) = length($line);
        #
        # extract the first word as a tag.
        #
	my($offset) = $column;
	while ($offset > 0 && substr($line, $offset, 1) =~ /^\w/) {
	    $offset--;
	}
	while ($offset < $length && substr($line, $offset, 1) !~ /^\w/) {
	    $offset++;
	}
	my($subline) = substr($line, $offset);
        ($tag, $blace) = ($subline =~ /(\w+)\s*(\(?)/);
        if (!$tag) {
            $curscr->Msg("tag not found in current position.");
            return;
        }
        #
        # decide flag value.
        #
	$flag = 's';
        if ($offset == 0) {
	    if ($blace) {
                $flag = 'r';		# maybe function definition.
	    }
        } else {
	    if ($line =~ /^#\s*define\s+(\w+)\(/ && $1 eq $tag) {
                $flag = 'r';		# maybe macro definition.
	    } elsif ($line =~ /^(ENTRY|ALTENTRY|NENTRY)\((\w+)\)$/ && $2 eq $tag) {
                $flag = 'r';		# maybe assember function definition.
	    } elsif ($blace) {
                $flag = 'x';		# maybe function reference.
	    }
        }
    }
    open(TAGS, "$command -xq$flag '$tag' |");
    $tagq = undef;
    while(<TAGS>) {
        my ($name, $lno, $path, $rest);
        if ((($name, $lno, $path, $rest) = /^([^ \t]+)[ \t]+(\d+)[ \t]+([^ \t]+)(.*)$/) >= 2) {
	    if (!$tagq) {
		$tagq = $curscr->TagQ($tag);
	    }
           $tagq->Add($path, $lno, '');
        }
    } 
    close(TAGS);
    $status = $?;
    $status = $status / 256;
    if ($status == 0) {
        if (!$tagq) {
	    $curscr->Msg("$tag: tag not found.");
        } else {
	    $tagq->Push();
        }
    } elsif ($status == 1) {
        $curscr->Msg("Global(1) failed. Please test it out of nvi.");
    } elsif ($status == 2) {
        $curscr->Msg("Usage error. Please type ':perl help[ENTER]'.");
    } elsif ($status == 3) {
        $curscr->Msg("GTAGS not found.");
    } else {
        $curscr->Msg("Unknown error.");
    }
}
sub main::gozilla {
    if ($notfound) {
        $curscr->Msg($notfound);
	return;
    }
    my($filename) = $curscr->GetFileName();
    my($lineno, $column) = $curscr->GetCursor();
    system("gozilla +$lineno $filename");
}
sub main::help {
    $help = <<END_OF_HELP;
Commands:
-f files: Print all function definitions in the files.
-g pattern: Print all lines which match to the pattern.
-I pattern: Print all lines which match to the pattern.
-P pattern: Print the path which match to the pattern.
Options:
-i: ignore case distinctions in pattern.
-l: Print just objects which exist under the current directory.
-r: Print the locations of object references.
-s: Print the locations of specified symbol other than function names.
-T: Go through all the tag files listed in GTAGSLIBPATH.
END_OF_HELP
    $curscr->Msg($help);    
}

1;
